fix: insert user html more consistently

ojeytonwilliams opened the pull request:
## The issue
One of the problems we've had while writing challenges that inspect the DOM is that everything ends up in the `body` element, even if a user explicitly writes it in the `head`

To reproduce:

1. Go to https://www.freecodecamp.org/learn/responsive-web-design/basic-html-and-html5/say-hello-to-html-elements
2. paste the following into the editor
```html
<head><title>head title</title></head>
<body><title>body title</title></body>
<h1>Hello World</h1>
```
3. inspect the preview pane
4. see (link, script and comments removed for clarity)
```html
<html>
<head></head>
<body id="display-body" style="margin:8px;">
<title>head title</title>
<title>body title</title>
<h1>Hello World</h1>
</body>
</html>
```
5. create an index.html file and insert the original code, you will see:
```html
<html>
<head>
<title>head title</title>
</head>
<body>
<title>body title</title>
<h1>Hello World</h1>
</body>
</html>
```

As a result, we can't tell where `<title>head title</title>` was, it will end up in the body regardless.

## What this PR fixes

By using an `iframe` to parse the user's HTML (which needs to happen for SASS and Babel to work) we can overwrite a `html` element without wiping the entire page.  This preserves the user's `head` tags.

## What this PR doesn't fix

`html` tags are still ignored.  It doesn't seem to be possible to replace the root element entirely. `replaceWith(newHtml)` and similar methods either fail entirely or are no better than replacing the `innerHTML`.

## Why is this a draft?

Since we're not pushing everything into the body, the inline style that sets the body margin to `8px` had to be removed.  We're still discussing what to do about normalise and the inline style was used to restore the Chrome default of `8px`.

ojeytonwilliams committed:
fix: use an iframe to preserve head and body

ojeytonwilliams committed:
fix: remove unnecessary parsing of html

The contents gets inserted into the DOM during transformHtml, which
is always part of the build pipeline

ShaunSHamilton commented:
Oliver, thank you, for this PR. It will do well to help the tests.
What this PR does well:
- Better mimicks browser behaviour
- Allows tests to be written sans RegEx, knowing whether the `title` has been placed in the body or not

Implications:
- The `title` can still be placed outside the `body`, and end up in the `head` - browser behaviour.

ojeytonwilliams commented:
> The title can still be placed outside the body, and end up in the head - browser behaviour.

Is this just the browser fixing bad HTML in the usual way or is it peculiar to our implementation?

ShaunSHamilton commented:
> > The title can still be placed outside the body, and end up in the head - browser behaviour.
> 
> Is this just the browser fixing bad HTML in the usual way or is it peculiar to our implementation?

This is just browser being browser, and fixing the HTML for us.

raisedadead commented:
Cleaning things up. Added this to your Trello, so that you can come back and resurrect this when you are ready.

ojeytonwilliams commented:
@moT01 ran into some issues (margins in the iframe) that this clears up, and we're much closer to launching the front 3, so I think this should be revisited.

ojeytonwilliams commented:
@moT01 @ShaunSHamilton @nhcarrigan 

The removal of the margin looks pretty okay, honestly.  I checked a few challenges and it is noticeable if you're looking for it, but I don't think it's too bad.  Take a look at some of the early HTML challenges and let me know what you think.

ojeytonwilliams committed:
fix: pipe contents through iframe

ojeytonwilliams committed:
refactor: use the same code for both transforms

moT01 reviewed (dismissed):
Assuming tests pass, this LGTM :tada: 

I looked at quite a few challenges across every section that uses the preview, I'm convinced that this is the way to go. I didn't see any instances where it made enough of a difference that I would say not to do this.

ojeytonwilliams committed:
fix: try to handle test errors better

ojeytonwilliams committed:
DEBUG: increasing the test-challenges timeout

The existing timeout should have been plenty, but this bumps it up a
bit to confirm that the underlying problem really is the tests taking
too long.

ojeytonwilliams committed:
DEBUG: removing possible memory leak

ojeytonwilliams committed:
DEBUG: reset all the things

Resetting both JSDOM and puppeteer and hoping for the best

ojeytonwilliams committed:
test: report more info on unhandled rejections

ojeytonwilliams committed:
Revert "DEBUG: removing possible memory leak"

This reverts commit d25e5b390f1495b8a7989d159512d97392c7fe74.

moT01 committed:
fix: it?

moT01 committed:
test: revert test-challenges.js

ShaunSHamilton commented:
Potentially relevant: https://github.com/jsdom/jsdom/blob/master/README.md#basic-options

>By default, jsdom will not load any subresources such as scripts, stylesheets, images, or iframes. If you'd like jsdom to load such resources, you can pass the resources: "usable" option, which will load all usable resources.

moT01 committed:
he's heating up

moT01 committed:
why not

moT01 committed:
:sad-face:

moT01 committed:
again

moT01 committed:
last try

ojeytonwilliams commented:
It's been a while, so I took another look. Everything passed locally, but I'm trying it again in CI.

ojeytonwilliams committed:
test: update steps to look inside head, not body

ShaunSHamilton commented on curriculum/test/test-challenges.js:
```diff
@@ -88,7 +88,7 @@ const handleRejection = err => {
 };
 
-const dom = new jsdom.JSDOM('');
+const dom = new jsdom.JSDOM('', { resources: 'usable' });
```
Did this fix the issue, or was the overflow really caused by the tests searching the wrong element?

ojeytonwilliams replied:
Nothing has changed since November, so I think it fixed itself.  I'm going to revert all the 'fixes' see if it's still fine.

ShaunSHamilton reviewed (dismissed):
LGTM 👍

ojeytonwilliams committed:
chore: revert some fixes

ojeytonwilliams committed:
test: restore timeout

ojeytonwilliams committed:
chore: remove useless comment

moT01 reviewed (approved):
LGTM 🎉

ShaunSHamilton reviewed (approved):
LGTM 👍 

---

There are now useless comments in the curriculum challenge tests, because they mention this. That is unrelated to us getting this in, though.

Thank you, Oliver ⭐

ojeytonwilliams commented:
@ShaunSHamilton you're very welcome.  I hope it makes regex a fair bit less necessary.
